File: backend_snippet/goals_calculator.py
"""
Research-based daily targets calculator (Mifflin-St Jeor BMR, TDEE).
Use this in your Django backend so PATCH profile updates and persists daily_targets.
"""

# Activity multipliers for TDEE (TDEE = BMR × multiplier)
ACTIVITY_MULTIPLIERS = {
    "sedentary": 1.2,
    "light": 1.375,
    "moderate": 1.55,
    "very_active": 1.725,
    "extra_active": 1.9,
}


def bmr(weight_kg: float, height_cm: float, age: int, is_female: bool) -> float:
    """Mifflin-St Jeor BMR (kcal/day)."""
    base = (10 * weight_kg) + (6.25 * height_cm) - (5 * age)
    return base + (-161 if is_female else 5)


def tdee(
    weight_kg: float,
    height_cm: float,
    age: int,
    activity_level: str,
    is_female: bool = False,
) -> float:
    """Total Daily Energy Expenditure (kcal/day)."""
    bmr_val = bmr(weight_kg, height_cm, age, is_female)
    mult = ACTIVITY_MULTIPLIERS.get(
        activity_level.strip().lower().replace(" ", "_"), 1.55
    )
    return bmr_val * mult


def target_calories(tdee_value: float, weight_goal: str) -> float:
    """Target calories from TDEE and weight goal."""
    goal = weight_goal.strip().lower()
    if goal == "lose":
        return max(1200.0, tdee_value - 500)
    if goal == "gain":
        return tdee_value + 300
    return tdee_value


def target_macros(
    calories: float,
    weight_kg: float,
    weight_goal: str,
) -> dict:
    """Return dict with calories, protein_g, carbs_g, fat_g (research-based split)."""
    goal = weight_goal.strip().lower()
    protein_per_kg = 1.8 if goal in ("lose", "gain") else 1.2
    protein_g = max(50, min(300, weight_kg * protein_per_kg))
    protein_kcal = protein_g * 4
    fat_pct = 0.25
    fat_kcal = calories * fat_pct
    fat_g = fat_kcal / 9
    carbs_kcal = calories - protein_kcal - fat_kcal
    carbs_g = max(0, min(500, carbs_kcal / 4))
    return {
        "calories": round(calories),
        "protein_g": round(protein_g, 1),
        "carbs_g": round(carbs_g, 1),
        "fat_g": round(fat_g, 1),
    }


def compute_daily_targets(
    weight_kg: float,
    height_cm: float,
    age: int,
    activity_level: str,
    weight_goal: str,
    is_female: bool = False,
) -> dict:
    """Compute daily_targets from profile fields (same logic as Flutter app)."""
    tdee_val = tdee(weight_kg, height_cm, age, activity_level, is_female)
    cal = target_calories(tdee_val, weight_goal)
    return target_macros(cal, weight_kg, weight_goal)

File: lib/core/goals_calculator.dart
/// Research-based goal calculation (Mifflin-St Jeor BMR, TDEE, macro split).
/// Used when API profile has no daily_targets or as fallback.

class GoalsCalculator {
  /// Activity multipliers (TDEE = BMR × multiplier)
  static const Map<String, double> activityMultipliers = {
    'sedentary': 1.2,
    'lightly_active': 1.375,
    'moderate': 1.55,
    'very_active': 1.725,
    'extra_active': 1.9,
  };

  /// Mifflin-St Jeor BMR (kcal/day)
  /// Men: (10 × weight_kg) + (6.25 × height_cm) − (5 × age) + 5
  /// Women: (10 × weight_kg) + (6.25 × height_cm) − (5 × age) − 161
  static double bmr({
    required num weightKg,
    required num heightCm,
    required int age,
    bool isFemale = false,
  }) {
    final w = weightKg.toDouble();
    final h = heightCm.toDouble();
    final base = (10 * w) + (6.25 * h) - (5 * age);
    return base + (isFemale ? -161 : 5);
  }

  /// TDEE (Total Daily Energy Expenditure) in kcal/day
  static double tdee({
    required num weightKg,
    required num heightCm,
    required int age,
    required String activityLevel,
    bool isFemale = false,
  }) {
    final b = bmr(
      weightKg: weightKg,
      heightCm: heightCm,
      age: age,
      isFemale: isFemale,
    );
    final mult =
        activityMultipliers[activityLevel.toLowerCase().replaceAll(' ', '_')] ??
        1.55;
    return b * mult;
  }

  /// Target calories based on weight goal (lose / maintain / gain)
  /// Lose: -500 kcal/day (~0.5 kg/week), Gain: +300 kcal/day
  static double targetCalories({
    required double tdeeValue,
    required String weightGoal,
  }) {
    switch (weightGoal.toLowerCase()) {
      case 'lose':
        return (tdeeValue - 500).clamp(1200.0, double.infinity);
      case 'gain':
        return tdeeValue + 300;
      case 'maintain':
      default:
        return tdeeValue;
    }
  }

  /// Macro split: protein 30%, carbs 45%, fat 25% (common research-based default).
  /// Protein at least 1.6 g/kg for weight loss, 1.2 for maintain.
  static Map<String, double> targetMacros({
    required double calories,
    required num weightKg,
    required String weightGoal,
    double proteinPct = 0.30,
    double carbsPct = 0.45,
    double fatPct = 0.25,
  }) {
    final w = weightKg.toDouble();
    // Protein: at least 1.6 g/kg for lose, 1.2 for maintain, 1.6 for gain
    final proteinPerKg = weightGoal.toLowerCase() == 'lose'
        ? 1.8
        : (weightGoal.toLowerCase() == 'gain' ? 1.8 : 1.2);
    final proteinG = (w * proteinPerKg).clamp(50.0, 300.0);
    final proteinKcal = proteinG * 4;
    final remainingKcal = calories - proteinKcal;
    // Fat: 25% of total, min 0.5 g/kg
    final fatKcal = calories * fatPct;
    final fatG = fatKcal / 9;
    final carbsKcal = remainingKcal - fatKcal;
    final carbsG = (carbsKcal / 4).clamp(0.0, 500.0);
    return {
      'calories': calories.roundToDouble(),
      'protein_g': proteinG.roundToDouble(),
      'carbs_g': carbsG.roundToDouble(),
      'fat_g': fatG.roundToDouble(),
    };
  }

  /// Approximate kcal to lose or gain 1 kg body weight (research range ~7000–7800).
  static const double kcalPerKgBodyWeight = 7700;

  /// Estimated days to reach target weight at the given daily calorie target.
  /// Returns null for "maintain" or when estimate is not meaningful (e.g. no deficit/surplus).
  static int? estimatedDaysToTargetWeight({
    required num currentWeightKg,
    required num targetWeightKg,
    required double targetCaloriesPerDay,
    required double tdeeValue,
    required String weightGoal,
  }) {
    final goal = weightGoal.toLowerCase();
    if (goal == 'maintain') return null;

    if (goal == 'lose') {
      final kgToLose = currentWeightKg.toDouble() - targetWeightKg.toDouble();
      if (kgToLose <= 0) return null;
      final dailyDeficit = tdeeValue - targetCaloriesPerDay;
      if (dailyDeficit <= 0) return null;
      final days = (kgToLose * kcalPerKgBodyWeight) / dailyDeficit;
      return days.round().clamp(1, 365 * 2); // cap for display sanity
    }

    // gain
    final kgToGain = targetWeightKg.toDouble() - currentWeightKg.toDouble();
    if (kgToGain <= 0) return null;
    final dailySurplus = targetCaloriesPerDay - tdeeValue;
    if (dailySurplus <= 0) return null;
    final days = (kgToGain * kcalPerKgBodyWeight) / dailySurplus;
    return days.round().clamp(1, 365 * 2);
  }

  /// Full calculation: TDEE → target calories → macros
  static Map<String, double> calculateDailyTargets({
    required num weightKg,
    required num heightCm,
    required int age,
    required String activityLevel,
    required String weightGoal,
    bool isFemale = false,
  }) {
    final tdeeVal = tdee(
      weightKg: weightKg,
      heightCm: heightCm,
      age: age,
      activityLevel: activityLevel,
      isFemale: isFemale,
    );
    final cal = targetCalories(tdeeValue: tdeeVal, weightGoal: weightGoal);
    return targetMacros(
      calories: cal,
      weightKg: weightKg,
      weightGoal: weightGoal,
    );
  }
}
